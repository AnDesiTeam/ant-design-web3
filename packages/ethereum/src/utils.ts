import debug from 'debug';

const createDebug = (namespace: string) => {
  return debug(`antd-web3:ethereum:${namespace}`);
};

export { createDebug };
