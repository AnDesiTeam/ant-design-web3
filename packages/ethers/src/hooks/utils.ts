import * as ethers from 'ethers';
import * as ethers5 from 'ethers5';

export const isEthersV5 = () => ethers.version && ethers.version.startsWith('5.');

export const FallbackProvider = isEthersV5()
  ? ethers5.providers.FallbackProvider
  : ethers.FallbackProvider;

export const JsonRpcProvider = isEthersV5()
  ? ethers5.providers.JsonRpcProvider
  : ethers.JsonRpcProvider;

export const BrowserProvider = isEthersV5()
  ? ethers5.providers.Web3Provider
  : ethers.BrowserProvider;
