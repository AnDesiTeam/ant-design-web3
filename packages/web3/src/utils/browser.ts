export const getPlatform = () => {
  const userAgent = navigator.userAgent.toLowerCase();
  if (userAgent.includes('chrome')) {
    return 'Chrome';
  } else if (userAgent.includes('firefox')) {
    return 'Firefox';
  } else if (userAgent.includes('edge')) {
    return 'Edge';
  } else if (userAgent.includes('safari')) {
    return 'Safari';
  } else {
    return 'Other';
  }
};

export const writeCopyText = (text: string): Promise<void> => {
  return navigator.clipboard.writeText(text);
};

export const readCopyText = (): Promise<string> => {
  return navigator.clipboard.readText();
};
