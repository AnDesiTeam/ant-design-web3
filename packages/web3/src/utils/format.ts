export const fillWith0x = (address: string = ''): string => {
  const filledAddress = address.startsWith('0x') ? address : `0x${address}`;
  return filledAddress;
};

export const formatAddress = (address: string = '', groupSize = 4): string => {
  const formattedGroups = [];

  const has0x = address.startsWith('0x');
  const formatText = has0x ? address.slice(2) : address;

  for (let i = 0; i < formatText.length; i += groupSize) {
    const group = formatText.slice(i, i + groupSize);
    formattedGroups.push(group);
  }

  const formattedText = formattedGroups.join(' ');

  return has0x ? `0x ${formattedText}` : formattedText;
};

export const formatBalance = (value: bigint | number, decimals: number, fixed?: number): string => {
  const bigValue = typeof value === 'bigint' ? value : BigInt(value);
  const divisor = BigInt(10 ** decimals);
  const eth = bigValue / divisor;
  const ethFraction = bigValue % divisor;

  if (ethFraction === 0n && fixed === undefined) {
    return `${eth}`;
  }

  let fractionStr = ethFraction.toString().padStart(decimals, '0');
  if (fixed === undefined) {
    return `${eth}.${fractionStr.replace(/0+$/, '')}`;
  }
  fractionStr = fractionStr.substring(0, fixed).padEnd(fixed, '0');
  return `${eth}.${fractionStr}`;
};
