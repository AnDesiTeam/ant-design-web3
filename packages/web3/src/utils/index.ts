export const getPlatform = () => {
  const userAgent = navigator.userAgent.toLowerCase();

  if (userAgent.includes('chrome')) {
    return 'Chrome';
  } else if (userAgent.includes('firefox')) {
    return 'Firefox';
  } else if (userAgent.includes('edge')) {
    return 'Edge';
  } else if (userAgent.includes('safari')) {
    return 'Safari';
  } else {
    return 'Other';
  }
};
