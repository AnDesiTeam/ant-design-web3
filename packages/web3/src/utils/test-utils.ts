import { vi } from 'vitest';

const oriClipboard = window.navigator.clipboard;
// 代理 navigator.clipboard 方法用于测试复制文本到粘贴板的功能
export const mockClipboard = () => {
  const clipboard = {
    text: '',
    writeText: (text: string) => {
      clipboard.text = text;
      return Promise.resolve();
    },
    readText: () => {
      return Promise.resolve(clipboard.text);
    },
  };
  Object.defineProperty(window, 'navigator', {
    value: {
      clipboard,
    },
  });

  return () => {
    Object.defineProperty(window, 'navigator', {
      value: {
        clipboard: oriClipboard,
      },
    });
  };
};

/**
 * Wait for a time delay. Will wait `advanceTime * times` ms.
 *
 * @param advanceTime Default 1000
 * @param times Default 20
 */
export async function waitFakeTimer(advanceTime = 1000, times = 20) {
  for (let i = 0; i < times; i += 1) {
    // eslint-disable-next-line no-await-in-loop
    await vi.waitFor(async () => {
      await Promise.resolve();

      if (advanceTime > 0) {
        vi.advanceTimersByTime(advanceTime);
      } else {
        vi.runAllTimers();
      }
    });
  }
}
