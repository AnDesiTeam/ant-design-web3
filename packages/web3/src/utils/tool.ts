import type { GlobalToken } from 'antd';
import { TinyColor } from '@ctrl/tinycolor';

export const isDarkTheme = (token: GlobalToken) => {
  const hsv = new TinyColor(token.colorFill).onBackground(token.colorBgElevated).toHsv();
  return hsv.v < 0.5;
};

export const formatNumUnit = (num: number) => {
  return num > 1000 ? `${(num / 1000).toFixed(1)}k` : num;
};
