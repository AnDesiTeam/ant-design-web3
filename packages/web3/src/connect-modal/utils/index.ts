import type React from 'react';
import type { Wallet } from '../interface';
import { getPlatform } from '../../utils';

export const mergeReactNodeProps = (
  node: React.ReactNode,
  defaultNode: React.ReactNode,
  resolveString?: (node: string) => React.ReactNode,
) => {
  if (node === null || node === false) return node;
  if (typeof node === 'string' && resolveString) return resolveString(node);
  return node ?? defaultNode;
};

export const defaultGroupOrder = (a: string, b: string) => a.localeCompare(b);

export const getWalletRoute = (wallet: Wallet) => {
  const curExtensions =
    wallet?.extensions && wallet.extensions.some((item) => item.key === getPlatform());

  if (curExtensions) {
    return 'wallet';
  }

  if (wallet?.app) {
    return 'qrCode';
  }

  return 'unknown';
};
