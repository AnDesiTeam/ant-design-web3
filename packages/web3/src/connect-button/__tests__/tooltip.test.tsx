import { ConnectButton } from '..';
import { fireEvent, render } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { mockClipboard } from '../../utils/test-utils';
import { readCopyText } from '../../utils';

describe('ConnectButton', () => {
  let resetMockClipboard: () => void;
  beforeEach(() => {
    resetMockClipboard = mockClipboard();
  });
  afterEach(() => {
    resetMockClipboard();
  });

  it('display tooltip', () => {
    const { baseElement } = render(
      <ConnectButton
        account={{
          address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18',
        }}
        tooltip={{ open: true }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      '0x3ea2cfd153b8d8505097b81c87c11f5d05097c18',
    );
    expect(baseElement.querySelector('.anticon-copy')).not.toBeNull();
  });

  it('disabled copyable in tooltip', () => {
    const { baseElement } = render(
      <ConnectButton
        account={{ address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18' }}
        tooltip={{ open: true, copyable: false }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      '0x3ea2cfd153b8d8505097b81c87c11f5d05097c18',
    );
    expect(baseElement.querySelector('.anticon-copy')).toBeNull();
  });

  it('custom title in tooltip', () => {
    const { baseElement } = render(
      <ConnectButton
        account={{ address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18' }}
        tooltip={{ open: true, title: 'aaaaaabbbbbbcccccc' }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      'aaaaaabbbbbbcccccc',
    );
  });

  it('should not display tooltip when not custom title and without address in tooltip', () => {
    const { baseElement } = render(<ConnectButton tooltip />);
    expect(baseElement.querySelector('.ant-tooltip')).toBeNull();
  });

  it('should copy text after click copy icon', async () => {
    const { baseElement } = render(
      <ConnectButton
        account={{ address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18' }}
        tooltip={{ open: true }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      '0x3ea2cfd153b8d8505097b81c87c11f5d05097c18',
    );
    expect(baseElement.querySelector('.anticon-copy')).not.toBeNull();
    fireEvent.click(baseElement.querySelector('.anticon-copy')!);
    await vi.waitFor(() => {
      expect(baseElement.querySelector('.ant-message')).not.toBeNull();
      expect(baseElement.querySelector('.ant-message-notice-content')?.textContent).toBe(
        'Address Copied!',
      );
      expect(readCopyText()).resolves.toBe('0x3ea2cfd153b8d8505097b81c87c11f5d05097c18');
    });
  });

  it('should copy text after click copy icon in custom title mode', async () => {
    const { baseElement } = render(
      <ConnectButton tooltip={{ open: true, title: 'aaaaaabbbbbbcccccc' }} />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      'aaaaaabbbbbbcccccc',
    );
    expect(baseElement.querySelector('.anticon-copy')).not.toBeNull();
    fireEvent.click(baseElement.querySelector('.anticon-copy')!);
    await vi.waitFor(() => {
      expect(baseElement.querySelector('.ant-message')).not.toBeNull();
      expect(baseElement.querySelector('.ant-message-notice-content')?.textContent?.trim()).toBe(
        'Address Copied!',
      );
      expect(readCopyText()).resolves.toBe('aaaaaabbbbbbcccccc');
    });
  });
});
