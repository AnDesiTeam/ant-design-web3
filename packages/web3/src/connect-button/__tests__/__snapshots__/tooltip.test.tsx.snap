// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConnectButton > should display formatted by custom formatter when pass format into tooltip 1`] = `
<body>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-1ck3jst ant-btn-default css-dev-only-do-not-override-1ck3jst ant-web3-connect-button ant-tooltip-open"
      type="button"
    >
      <div
        class="ant-web3-connect-button-text css-dev-only-do-not-override-1ck3jst"
      >
        <div
          class="ant-space css-dev-only-do-not-override-1ck3jst ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small ant-web3-address css-dev-only-do-not-override-1ck3jst"
        >
          <div
            class="ant-space-item"
          >
            <span
              class="ant-web3-address-text"
            >
              0x3ea2...7c18
            </span>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div>
    <div
      class="ant-tooltip ant-zoom-big-fast-appear ant-zoom-big-fast-appear-prepare ant-zoom-big-fast ant-web3-connect-button-tooltip css-dev-only-do-not-override-1ck3jst css-dev-only-do-not-override-1ck3jst ant-tooltip-placement-top"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
    >
      <div
        class="ant-tooltip-arrow"
        style="position: absolute; bottom: 0px; left: 0px;"
      />
      <div
        class="ant-tooltip-content"
      >
        <div
          class="ant-tooltip-inner"
          role="tooltip"
        >
          <div
            class="ant-web3-connect-button-tooltip-title"
          >
            Wallet address
             
            <span
              aria-label="copy"
              class="anticon anticon-copy"
              role="img"
              tabindex="-1"
              title="Copy Address"
            >
              <svg
                aria-hidden="true"
                data-icon="copy"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                />
              </svg>
            </span>
          </div>
          <div
            class="ant-web3-connect-button-tooltip-content"
          >
            0x3ea2cfd1
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ConnectButton > should display formatted when pass format into tooltip 1`] = `
<body>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-1ck3jst ant-btn-default css-dev-only-do-not-override-1ck3jst ant-web3-connect-button ant-tooltip-open"
      type="button"
    >
      <div
        class="ant-web3-connect-button-text css-dev-only-do-not-override-1ck3jst"
      >
        <div
          class="ant-space css-dev-only-do-not-override-1ck3jst ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small ant-web3-address css-dev-only-do-not-override-1ck3jst"
        >
          <div
            class="ant-space-item"
          >
            <span
              class="ant-web3-address-text"
            >
              0x3ea2...7c18
            </span>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div>
    <div
      class="ant-tooltip ant-zoom-big-fast-appear ant-zoom-big-fast-appear-start ant-zoom-big-fast ant-web3-connect-button-tooltip css-dev-only-do-not-override-1ck3jst css-dev-only-do-not-override-1ck3jst ant-tooltip-placement-top"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box; transition: none;"
    >
      <div
        class="ant-tooltip-arrow"
        style="position: absolute; bottom: 0px; left: 0px;"
      />
      <div
        class="ant-tooltip-content"
      >
        <div
          class="ant-tooltip-inner"
          role="tooltip"
        >
          <div
            class="ant-web3-connect-button-tooltip-title"
          >
            Wallet address
             
            <span
              aria-label="copy"
              class="anticon anticon-copy"
              role="img"
              tabindex="-1"
              title="Copy Address"
            >
              <svg
                aria-hidden="true"
                data-icon="copy"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                />
              </svg>
            </span>
          </div>
          <div
            class="ant-web3-connect-button-tooltip-content"
          >
            0x 3ea2 cfd1 53b8 d850 5097 b81c 87c1 1f5d 0509 7c18
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-message ant-message-top css-dev-only-do-not-override-1ck3jst"
    style="left: 50%; transform: translateX(-50%); top: 8px;"
  >
    <div
      class="ant-message-notice ant-message-move-up-appear ant-message-move-up-appear-active ant-message-move-up ant-message-notice-success css-dev-only-do-not-override-1ck3jst"
    >
      <div
        class="ant-message-notice-content"
      >
        <div
          class="ant-message-custom-content ant-message-success"
        >
          <span
            aria-label="check-circle"
            class="anticon anticon-check-circle"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="check-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 01-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
              />
            </svg>
          </span>
          <span>
            Address Copied!
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
